libkbfs/folder_block_ops.go

```diff
@@ -975,54 +1065,78 @@ func (fbo *folderBlockOps) setCachedAttrLocked(
 // up the cache entry when the effects of the operation are no longer
 // needed.
 func (fbo *folderBlockOps) SetAttrInDirEntryInCache(lState *lockState,
-	p path, newDe DirEntry, attr attrChange) (deleteTargetDirEntry bool) {
+	p path, newDe DirEntry, attr attrChange) dirCacheUndoFn {
 	fbo.blockLock.Lock(lState)
 	defer fbo.blockLock.Unlock(lState)
 
 	// Pretend this is a directory add, to ensure the directory is
 	// synced.
-	fbo.addDirEntryInCacheLocked(lState, *p.parentPath(), p.tailName(), newDe)
+	undoAdd := fbo.addDirEntryInCacheLocked(
+		lState, *p.parentPath(), p.tailName(), newDe)
 
+	// TODO(KBFS-2076): Uncomment below (see comment in undo function).
 	// Update the actual attribute in the deCache.
+	// cacheEntry, ok := fbo.deCache[newDe.Ref()]
 	_, ok := fbo.deCache[newDe.Ref()]
-	if !ok {
-		deleteTargetDirEntry = true
-	}
+	// cacheEntryCopy := cacheEntry.deepCopy()
 	fbo.setCachedAttrLocked(
 		lState, newDe.Ref(), attr, &newDe,
 		true /* create the deCache entry if it doesn't exist yet */)
-	return deleteTargetDirEntry
+	return fbo.wrapWithBlockLock(func() {
+		if ok {
+			// TODO(KBFS-2076): Uncomment below.  Before KBFS-2076,
+			// the undo shouldn't restore attributes that are set in
+			// an entry that was created by a write, since it is
+			// always called even after a successfully-synced setattr
+			// call.
+			//fbo.deCache[newDe.Ref()] = cacheEntryCopy
+		} else {
+			delete(fbo.deCache, newDe.Ref())
+		}
+		undoAdd()
+	})
 }
 
 // ClearCachedAddsAndRemoves clears out any cached directory entry
 // adds and removes for the given dir.
-func (fbo *folderBlockOps) ClearCachedAddsAndRemoves(
-	lState *lockState, dir path) {
+func (fbo *folderBlockOps) ClearCachedAdd(
+	lState *lockState, dir path, name string) {
 	fbo.blockLock.Lock(lState)
 	defer fbo.blockLock.Unlock(lState)
 	cacheEntry, ok := fbo.deCache[dir.tailPointer().Ref()]
 	if !ok {
 		return
 	}
 
-	// If there's no dirEntry, we can just delete the whole thing.
-	if !cacheEntry.dirEntry.IsInitialized() {
+	delete(cacheEntry.adds, name)
```

Comment:
Interesting that you can "delete" from a nil map, but not assign to it. No action here, just thought it was interesting.